[21710] Hook for remote managed repositories

oliverguenther opened the pull request:
This PR provides functionality to notify with remote services for managing repositories.

The main driver behind this is packager, for which we want apache to be the sole owner of the repositories it delivers, and OpenProject only as a reader, however it benefits other installations as well.

Relevant work package: https://community.openproject.org/work_packages/21710/activity
Actual apache wrapper contained in: https://github.com/opf/openproject/pull/3631

oliverguenther committed:
Remote managed repositories

This commit starts the work to provide functionality with remote service hooks for managing repositories.

The main driver behind this is packager, for which we want apache to be the sole owner of the repositories it delivers, and OpenProject only as a reader, however it benefits other installations as well.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-reposman-api.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-reposman-api.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

oliverguenther committed:
Extend exemplary configuration

oliverguenther committed:
Enable webmock as a context explictly in examples

As webmock breaks default Net::HTTP behavior, it should not be run unless explictly requested.
Requiring it in the spec_helper breaks some other integration tests.

oliverguenther committed:
Feature specs for remote repository management

Introduces a metadata key `:webmock` that can be used to enable
mocking HTTP requests in an environment on demand.

oliverguenther commented:
@opf/developers Need some feedback to get this merged so that I can complete packager-based repositories with https://github.com/opf/openproject/pull/3631

basiszwo commented on spec/services/scm/create_managed_repository_service_spec.rb:
```diff
@@ -134,4 +136,43 @@
   end
+
+  context 'with managed remote config' do
+    include_context 'with webmock'
```
you could use metadata as well. https://www.relishapp.com/rspec/rspec-core/docs/hooks/filters#filter-`around(:example)`-hooks-using-arbitrary-metadata

maybe more obvious than a context? (maybe just gusto ...)

oliverguenther replied:
Thinking about it, i would expect to receive some state from the context (e.g., a new `let` as in the tmpdir shared context).

I'll refactor it to metadata.

basiszwo commented on app/services/scm/delete_managed_repository_service.rb:
```diff
@@ -35,14 +35,17 @@
+
+    if repository.class.manages_remote?
+      Scm::DeleteRemoteRepositoryJob.new(repository).perform
+      true
```
why do you always need to return true?

oliverguenther replied:
I just wanted the service be responsible to keep the result of `.call` a boolean instead of whatever the job returns. It should not make a difference in the end, but i prefer being more explicit here.

basiszwo commented on app/views/repositories/settings/shared/_managed.html.erb:
```diff
@@ -0,0 +1,56 @@
+      <%= form.text_field(:url,
+                          label: l('repositories.managed_url'),
+                          value: l('repositories.managed_remote'),
+                          size: 60,
```
is there a reason for a 60 character limit to the remote url?

oliverguenther replied:
I think that was simply historical. I'll remove it.

basiszwo commented:
Added some very minor feedback. Besides that this looks good to me :+1:
